### PR TITLE
Industry manager functions 2

### PR DESCRIPTION
**Pull request by duncanspumpkin**
wip

### CODE REVIEW

File: src/OpenLoco/IndustryManager.cpp [hunk 1]
```diff
@@ -200,15 +200,21 @@ namespace OpenLoco::IndustryManager
...
-        const auto shareLimit = share2 + share3;
+
+        // var_CE is in the range of 1->32 inclusive
+        // This formula is ultimately staticPreferredTotalOfType = 1/4 * ((getNumIndustries() + 1) * indObj->var_CE)
+        // but we will do it in two steps to keep identical results.
+        const auto intermediate1 = ((getNumIndustries() + 1) * indObj->var_CE) / 3;
+        const auto staticPreferredTotalOfType = intermediate1 - intermediate1 / 4;
+        // The preferred total can vary by up to a half of the static preffered total.
+        const auto randomPrefferedTotalOfType = (staticPreferredTotalOfType / 2) * gPrng().randNext(0xFF) / 256;
+        const auto prefferedTotalOfType = staticPreferredTotalOfType + randomPrefferedTotalOfType;
```
Comment (AaronVanGeffen) [inline src/OpenLoco/IndustryManager.cpp:211]:
```suggestion
        const auto randomPreferredTotalOfType = (staticPreferredTotalOfType / 2) * gPrng().randNext(0xFF) / 256;
        const auto preferredTotalOfType = staticPreferredTotalOfType + randomPreferredTotalOfType;
```

File: src/OpenLoco/IndustryManager.cpp [hunk 2]
```diff
@@ -200,15 +200,21 @@ namespace OpenLoco::IndustryManager
...
+
         const auto totalOfThisType = std::count_if(std::begin(industries()), std::end(industries()), [indObjId](const auto& industry) {
             return (industry.object_id == indObjId);
         });
 
-        if (totalOfThisType < shareLimit)
+        if (totalOfThisType < prefferedTotalOfType)
```
Comment (AaronVanGeffen) [inline src/OpenLoco/IndustryManager.cpp:217]:
```suggestion
        if (totalOfThisType < preferredTotalOfType)
```

File: src/OpenLoco/IndustryManager.cpp [hunk 3]
```diff
@@ -70,10 +78,205 @@ namespace OpenLoco::IndustryManager
         }
     }
 
+    // 0x0047EA42
+    static size_t getMostCommonBuildingCargoType()
+    {
+        // First generate a count of all the different cargo based on what building could generate
+        std::array<uint32_t, 32> cargoCounts{};
```
Comment (ZehMatt) [inline src/OpenLoco/IndustryManager.cpp:85]: Magic number here, this should be a constant somewhere.

**Comment by duncanspumpkin** [comment]
No divergence on 1000 ticks of the lets play save. Not sure if that is a great test though for this.

**Comment by AaronVanGeffen** [comment]
How many in-game days does 1000 ticks translate to? I'd suggest testing for at least one year in-game.

**Comment by LeftofZen** [comment]
what do we need to do for this one? more testing?

**Comment by duncanspumpkin** [comment]
> what do we need to do for this one? more testing?

Yes pretty much. I'm afk for the week so I'm unable to do it till next week

**Comment by LeftofZen** [comment]
Played normally for about 4 in-game years, couldn't see any issues

File: src/OpenLoco/Objects/IndustryObject.h [hunk 4]
```diff
@@ -52,10 +52,14 @@ namespace OpenLoco
...
-        uint8_t var_CE;
-        uint8_t cost_index;      // 0xCF
-        int16_t cost_factor;     // 0xD0
-        int16_t clearCostFactor; // 0xD2
+        // Total industries of this type that can be created in a scenario
+        // Note: this is not directly comparabile to total industries and vaires based
+        // on scenario total industries cap settings. At low industries cap this value is ~3x the
+        // amount of industries in a scenario.
+        uint8_t totalOfTypeInScenario; // 0xCE
+        uint8_t cost_index;            // 0xCF
```
Comment (LeftofZen) [inline src/OpenLoco/Objects/IndustryObject.h:60]: code style - `costIndex`, `costFactor`

**Review by LeftofZen — COMMENTED** [review]
LGTM, just fix the code style on the variables and its good

**Comment by AaronVanGeffen** [comment]
I would like for us to do some autosave comparisons for the usual save game as well. Perhaps as long as one year, in-game?

Ideally we'd test the emergence and disappearance of industries. as well, though, which the usual Let's Play save doesn't cover.

**Comment by duncanspumpkin** [comment]
No divergence on the 10000 ticks either (4 months). Does anyone have a save with lots of industry being created which I can test?

**Comment by LeftofZen** [comment]
I played this with the save @duncanspumpkin provided in discord (attached) and the game crashes on the first monthly update in Interop.cpp in the usual place
image


[1Industry.zip](https://github.com/OpenLoco/OpenLoco/files/9607652/1Industry.zip)

**Comment by duncanspumpkin** [comment]
> Member

woops. Fixed that crash.

**Comment by duncanspumpkin** [comment]
Getting a divergence on 10000 ticks of 1Industry.

**Comment by LeftofZen** [comment]
I didn't check divergence or anything, but the crashes are gone for me.

**Comment by duncanspumpkin** [comment]
No divergence on 10000 ticks of 1Industry.

**Pull request merged**